Userland/Libraries/LibCrypto/Curves/SECPxxxr1.h

```diff
@@ -0,0 +1,687 @@
+    {
+        // If condition = 0 return left else right
+        StorageType mask = static_cast<StorageType>(condition) - 1;
+        AK::taint_for_optimizer(mask);
+
+        return (left & mask) | (right & ~mask);
+    }
+
+    constexpr StorageType modular_reduce(StorageType const& value)
+    {
+        // Add -prime % 2^KEY_BIT_SIZE
+        bool carry = false;
+        StorageType other = value.addc(REDUCE_PRIME, carry);
+
+        // Check for overflow
+        return select(value, other, carry);
+    }
+
+    constexpr StorageType modular_reduce_order(StorageType const& value)
+    {
+        // Add -order % 2^KEY_BIT_SIZE
+        bool carry = false;
+        StorageType other = value.addc(REDUCE_ORDER, carry);
+
+        // Check for overflow
+        return select(value, other, carry);
+    }
+
+    constexpr StorageType modular_add(StorageType const& left, StorageType const& right, bool carry_in = false)
+    {
+        bool carry = carry_in;
+        StorageType output = left.addc(right, carry);
+
+        // If there is a carry, subtract p by adding 2^KEY_BIT_SIZE - p
+        StorageType addend = select(0u, REDUCE_PRIME, carry);
+        carry = false;
+        output = output.addc(addend, carry);
+
+        // If there is still a carry, subtract p by adding 2^KEY_BIT_SIZE - p
+        addend = select(0u, REDUCE_PRIME, carry);
+        return output + addend;
+    }
+
+    constexpr StorageType modular_sub(StorageType const& left, StorageType const& right)
+    {
+        bool borrow = false;
+        StorageType output = left.subc(right, borrow);
+
+        // If there is a borrow, add p by subtracting 2^KEY_BIT_SIZE - p
+        StorageType sub = select(0u, REDUCE_PRIME, borrow);
+        borrow = false;
+        output = output.subc(sub, borrow);
+
+        // If there is still a borrow, add p by subtracting 2^KEY_BIT_SIZE - p
+        sub = select(0u, REDUCE_PRIME, borrow);
+        return output - sub;
+    }
+
+    constexpr StorageType modular_multiply(StorageType const& left, StorageType const& right)
+    {
+        // Modular multiplication using the Montgomery method: https://en.wikipedia.org/wiki/Montgomery_modular_multiplication
+        // This requires that the inputs to this function are in Montgomery form.
+
+        // T = left * right
+        StorageTypeX2 mult = left.wide_multiply(right);
+        StorageType mult_mod_r = static_cast<StorageType>(mult);
+
+        // m = ((T mod R) * curve_p')
+        StorageType m = mult_mod_r * PRIME_INVERSE_MOD_R;
+
+        // mp = (m mod R) * curve_p
+        StorageTypeX2 mp = m.wide_multiply(PRIME);
+
+        // t = (T + mp)
+        bool carry = false;
+        mult_mod_r.addc(static_cast<StorageType>(mp), carry);
+
+        // output = t / R
+        StorageType mult_high = static_cast<StorageType>(mult >> KEY_BIT_SIZE);
+        StorageType mp_high = static_cast<StorageType>(mp >> KEY_BIT_SIZE);
```

Comment:
I must admit `low()` and `high()` here looked much better.

Reply:
I agree that `low()` and `high()` looked better, but I'm also thinking about implementing SECP521r1, where they will not work.

Reply:
Sure, ignore this then.

Re 521-bit integers: In case you haven't figured this out already, addc/subc won't work correctly for them. Overflow handling when `bit_size % 64 != 0` is a giant and unpleasant FIXME currently.

Reply:
Good to know, the SECP521r1 implementation is just all-round annoying as it does not even need an integer number of bytes.